### ISSUE TITLE
HPA with container metrics fails when any pod is not in a ready state

### ISSUE THREAD

**Issue by DP19** [kubernetes/kubernetes#127169]
### What happened?

We've noticed that our HPA will stay at it's current scale and not make decisions when a pod is in an un-Ready state. IE a single pod is in CrashLoopBackoff and the HPA shows:

```
...
resource cpu of container "mycontainer" on pods  (as a percentage of request):     <unknown> / 50%
...
Warning  FailedGetResourceMetric  78s (x7 over 4m7s)  horizontal-pod-autoscaler  unable to get metric cpu: failed to get container metrics: container guides3 not present in metrics for pod namesapce/mycontainer-20240904t1338-433ca6a1-db998df74-s5rfh
```

That particular pod was in CrashLoopBackoff at the time and it wasn't until i removed it was it able to make scaling decisions

We're also running istio so there is another container in these pods and it's running ok; this may be why it's showing up in the metrics but unable to get info on the main container

### What did you expect to happen?

I would expect the HPA to ignore pods in this state and use only the Ready pods for its decisions

### How can we reproduce it (as minimally and precisely as possible)?

Setup an HPA with container based metrics
Put one of the pods in bad state, could just have a single pod consume a lot of memory and have it OOM a few times
See the HPA make no decisions while it's in this state

### Anything else we need to know?

I've looked through the code and it seems to be steming from:
`metrics, err := c.client.PodMetricses(namespace).List(ctx, metav1.ListOptions{LabelSelector: selector.String()})`
This client just uses the selector and doesn't care about the state any of the pods might be in. It does end up filtering out pods later in the reconcile cycle but this returns an error here if it's not in the list of metrics because a single container may not be running 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.29.6
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.6-gke.1326000
```

</details>

### Cloud provider

<details>
GCP
</details>

**Comment by chengjoey** [comment]
The container metric should come from cadvisor. When an OOM occurs, the container will be deleted and recreated, so HPA may not be able to obtain the container's metrics at this time.

**Comment by DP19** [comment]
Yeah I understand _why_ it's unable to get metrics but this feels really bad to break the hpa behavior because of a single container; especially one that is in a CrashLoopBackOff state and could get exponentially longer in that state. 


I've seen other instances as well with similar behavior. Another container is on a node that is has memory pressure, gets soft evicted but continues to consume ram, node goes unresponsive and the hpa breaks in a similar way. This single pod not having metrics and in a Terminating state seems line something that should be ignored in favor of the Ready pods, which is what is considered by the end of the reconcile loop

**Comment by chengjoey** [comment]
Because this is horizontal expansion, if do not consider these CrashLoopBack pods, the so-called ready pods that are expanded are likely to become CrashLoopBack. Will this lead to the creation of more and more pods, and maybe CrashLoopBack still become ready after the container is rebuilt

If the container's metrics cannot be obtained and the container's status is OOM, is it feasible to treat its memory usage as 100%?

**Comment by DP19** [comment]
Sure that's half of the equation; in my situation it was scaled to max and unable to scale down even though all metrics of the running pods showed it well under the targets. But even in the Expansion, i'm not sure how adding more pods in a situation like that would cause them to become Crash Loop Back off if you're expanding the pool of pods to take on the workload?

> If the container's metrics cannot be obtained and the container's status is OOM, is it feasible to treat its memory usage as 100%?
Sure if that's what folks agree on. 

I would advocate for treating the returned value much like it does for the Pod metric:
After it gets the metrics in `GetMetricReplicas` we then pass it to `calcPlainMetricReplicas` which takes into consideration the state of the pods and removes them from the calculations 

```
	readyPodCount, unreadyPods, missingPods, ignoredPods := groupPods(podList, metrics, resource, c.cpuInitializationPeriod, c.delayOfInitialReadinessStatus)
	removeMetricsForPods(metrics, ignoredPods)
	removeMetricsForPods(metrics, unreadyPods)
```

instead it just errors out if it's not in the list of metrics found, bubbles up and ends the hpa decision cycle
```
		if !containerFound {
			return nil, fmt.Errorf("container %s not present in metrics for pod %s/%s", container, m.Namespace, m.Name)
		}
		
```


If nothing else there should be some doc update to let people know this is the behaviour and how it differs from the Pod metric based HPA calculations
